flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java

```diff
@@ -148,6 +162,33 @@ public Logger getLogger() {
         return logger;
     }
 
+    protected Throwable extractError(TaskExecutionStateTransition taskExecutionStateTransition) {
+        Throwable cause = taskExecutionStateTransition.getError(userCodeClassLoader);
+        if (cause == null) {
+            cause = new FlinkException("Unknown failure cause. Probably related to FLINK-21376.");
+        }
+        return cause;
+    }
+
+    protected Optional<ExecutionVertexID> extractExecutionVertexID(
+            TaskExecutionStateTransition taskExecutionStateTransition) {
+        return executionGraph.getExecutionVertexId(taskExecutionStateTransition.getID());
+    }
+
+    protected static Optional<RootExceptionHistoryEntry> convertFailures(
+            Function<ExecutionVertexID, Optional<ExecutionVertex>> lookup,
+            List<Failure> failureCollection) {
+        if (failureCollection.isEmpty()) {
+            return Optional.empty();
+        }
+        Failure first = failureCollection.remove(0);
+        Set<ExceptionHistoryEntry> entries = new HashSet<>();
+        for (Failure failure : failureCollection) {
+            entries.add(failure.toExceptionHistoryEntry(lookup));
```

Comment:
Do we want to filter redundant failures?
Do we want to scan all Executions for additional failures?

Reply:
Explanation:
**Filter redundant failures**
This can only happen, when a task is updated twice to failed with the same error. Or when the same global failure is reported twice.

**Scan all Executions**
Something like this:
`executions.filter(execution -> execution.getFailureInfo().isPresent()).collect(Collectors.toSet())`

This would identify executions that have failed, but not reported (or not in time) via `updateTaskExecutionState`.

Reply:
Wouldn't simply implementing hashCode & equals for the Failure object do the trick?

Reply:
I guess it could happen if TM doesn't get the ack response from the JM (some kind of transient network failure). But I'm not really sure whether TM retries than.

If it adds too much complexity I think it's OK not to bother with that. Maybe we should just use `List` instead of `Set` for collecting the entries to make the fact that we're not trying to be smart explicit.

Reply:
@zentol Do you know if `updateTaskExecutionState` could be retried by the TM? I guess we need to have something along these lines to make sure that the EG can transition to the terminal state.

Reply:
Your comment made me think. The concurrentExceptions are stored by `RootHistoryExceptionEntry` as a set. As we cannot (do not want to) change this, we need to enforce it. I will change the code accordingly.

Reply:
The `ExceptionHistoryEntry` doesn't properly implement hc & equals either, so there is no point of having set there either

Reply:
> Do you know if updateTaskExecutionState could be retried by the TM? I guess we need to have something along these lines to make sure that the EG can transition to the terminal state.

If the transition isn't confirmed the task fails on the TM, and the system will eventually reach a consistent state through the execution deployment reconciliation (failing the job).

Reply:
> The ExceptionHistoryEntry doesn't properly implement hc & equals either, so there is no point of having set there either

Good point. Fixed.